submissions/7SergSmith7/Tiny JS World/index.js

```diff
@@ -0,0 +1,115 @@
+/* Refer to https://github.com/OleksiyRudenko/a-tiny-JS-world for the task details
+   Complete the below for code reviewers' convenience:
+
+   Code repository: _put repo URL here_
+   Web app: _put project's github pages URL here_
+   */
+
+// ======== OBJECTS DEFINITIONS ========
+// Define your objects here
+class Habitant {
+  constructor(species, name, gender, phrase, friends = [], legs) {
+    this.species = species;
+    this.legs = legs;
+    this.name = name;
+    this.gender = gender;
+    this.phrase = phrase;
+    this.friends = friends;
+    this.addFriends(friends);
+
+    [
+      this.speciesProp,
+      this.nameProp,
+      this.genderProp,
+      this.phraseProp,
+      this.legsProp,
+    ] = ["species", "name", "gender", "phrase", "legs"].map(
+      (prop) => `<strong>${this[prop]}</strong>`
+    );
+  }
+
+  addFriends(friends) {
+    this.friends = [...this.friends, ...friends];
+  }
+
+  listFriends() {
+    return this.friends.length
+      ? this.friends.map((friend) => friend.name)
+      : "I am loner";
+  }
+
+  sayPhrase() {
+    return this.phrase;
+  }
+
+  toString() {
+    return `Hi! I am ${this.speciesProp}, my name is ${
+      this.nameProp
+    }, my gender is ${this.genderProp}.
+    My phrase: ${
+      this.phraseProp
+    } Friends: <strong>${this.listFriends()}</strong>. I have ${
+      this.legsProp
+    } legs `;
+  }
+}
+
+class Dog extends Habitant {
+  constructor(name, gender, phrase, friends, legs = 4) {
+    super("dog", name, gender, phrase, friends, legs);
+  }
+}
+class Cat extends Habitant {
+  constructor(name, gender, phrase, friends, legs = 4) {
+    super("cat", name, gender, phrase, friends, legs);
+  }
+}
+class Human extends Habitant {
+  constructor(name, gender, phrase, friends, legs = 2, hands = 2) {
+    super("human", name, gender, phrase, friends, legs, hands);
+    this.hands = hands;
+    [
+      this.speciesProp,
+      this.nameProp,
+      this.genderProp,
+      this.phraseProp,
+      this.legsProp,
+      this.handsProp,
+    ] = ["species", "name", "gender", "phrase", "legs", "hands"].map(
+      (prop) => `<strong>${this[prop]}</strong>`
+    );
```

Comment:
As `Human` extends `Habitant` the `Habitant`'s constructor gets called (actually thanks to the `super` call on the top of `Human`'s constructor. No need to construct `this.speciesProp` twice.

Besides this (and similar in the super class) fragment adds extra properties that are in fact only needed in `toString`.

Besides these `*Prop` props are carved in stone at construction time? What if we change any core prop like e.g. `this.hands` somewhere in the course of inhabitant's life and try to represent them? We will still have the "old" property value as output.

Suggested approach:
1. Create formatted values right where they are needed (i.e. in `toString`)
2. Do not store them as props but rather as temporary values every time the `toString` gets called.

Reply:
Done =)